### PR TITLE
llvm: replace some deprecated functions, add fixmes

### PR DESCRIPTION
**Pull request by klensy**
Replace some deprecated llvm functions, add FIXME's (for simpler future work), replace some rust custom functions with llvm ones.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs [hunk 1]
```diff
@@ -1072,7 +1072,13 @@ extern "C" {
     pub fn LLVMGetPoison(Ty: &Type) -> &Value;
 
     // Operations on metadata
+    // FIXME: deprecated, replace with LLVMMDStringInContext2
     pub fn LLVMMDStringInContext(C: &Context, Str: *const c_char, SLen: c_uint) -> &Value;
+
+    // LLVMMDStringInContext but don't own string
+    pub fn LLVMMDStringInContext2(C: &Context, Str: *const c_char, SLen: size_t) -> &Metadata;
```
Comment (klensy) [inline compiler/rustc_codegen_llvm/src/llvm/ffi.rs:1079]: I can probably replace `LLVMMDStringInContext` with `LLVMMDStringInContext2` all around, using value<->metadata wrappers around, but don't know is it worth it.

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs [hunk 2]
```diff
@@ -1290,34 +1298,34 @@ extern "C" {
...
 
-    pub fn LLVMRustBuildCleanupPad<'a>(
+    pub fn LLVMBuildCleanupPad<'a>(
         B: &Builder<'a>,
         ParentPad: Option<&'a Value>,
-        ArgCnt: c_uint,
         Args: *const &'a Value,
+        NumArgs: c_uint,
         Name: *const c_char,
     ) -> Option<&'a Value>;
```
Comment (klensy) [inline compiler/rustc_codegen_llvm/src/llvm/ffi.rs:1307]: Sometimes llvm type `LLVMValueRef` modeled in rust as `Option<&'a Value>`, sometimes as `&'a Value`, where `LLVMValueRef` can be nullptr:
https://github.com/llvm/llvm-project/blame/42d1b276f7793999be3f9b6a99efbb143254c729/llvm/lib/IR/Core.cpp#L3170-L3178

It feels wrong and can lead to funny results, or i'm wrong? Don't have strong c++ here, so advice requested.

**Review by WaffleLapkin — APPROVED** [review]

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs [hunk 3]
```diff
@@ -1072,7 +1072,13 @@ extern "C" {
     pub fn LLVMGetPoison(Ty: &Type) -> &Value;
 
     // Operations on metadata
+    // FIXME: deprecated, replace with LLVMMDStringInContext2
     pub fn LLVMMDStringInContext(C: &Context, Str: *const c_char, SLen: c_uint) -> &Value;
+
+    // LLVMMDStringInContext but don't own string
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/llvm/ffi.rs:1078]: The difference between these is whether the return value is Value or Metadata, nothing about ownership changes.

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs (hunk 1 above)
Comment (nikic) [reply compiler/rustc_codegen_llvm/src/llvm/ffi.rs:1079]: It looks like some of them are sensible to replace at least. E.g. in consts.rs LLVMMDStringInContext and LLVMMDNodeInContext can be replaced with the 2 variants (just need to replace both of them together and it works out fine).

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs [hunk 4]
```diff
@@ -1111,6 +1117,8 @@ extern "C" {
         Packed: Bool,
     ) -> &'a Value;
 
+    // FIXME: replace with LLVMConstArray2
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/llvm/ffi.rs:1120]: ... once our minimum version of LLVM 17.

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs (hunk 1 above)
Comment (klensy) [reply compiler/rustc_codegen_llvm/src/llvm/ffi.rs:1079]: Yes, but `void LLVMAddNamedMetadataOperand(LLVMModuleRef M, const char *Name, LLVMValueRef Val)` 
https://github.com/rust-lang/rust/blob/90a9f69c80812c8694959c1f2d5c336e3300d1e2/compiler/rustc_codegen_llvm/src/consts.rs#L507-L512
https://github.com/llvm/llvm-project/blame/02aa966c135172885182ddd6f5d76883c2402bed/llvm/lib/IR/Core.cpp#L1292-L1300
feeds in LLVMValueRef , but LLVMMDNodeInContext2 will return LLVMModuleRef, so wrap\unwrap (or rewrite LLVMAddNamedMetadataOperand).

Comment (klensy) [reply compiler/rustc_codegen_llvm/src/llvm/ffi.rs:1079]: Also `void LLVMSetMetadata(LLVMValueRef Inst, unsigned KindID, LLVMValueRef Val)` prevents refactor too, sadly.

**Comment by Nilstrieb** [comment]
maybe this was the cause of https://github.com/notifications#issuecomment-1500588281?
@bors rollup=iffy
feel free to r- if you think it's likely, but I'll leave it open for now to test it again

**Comment by klensy** [comment]
Yes, can reproduce when building stage2 compiler, but don't have idea why, for now.

**Pull request merged**